### PR TITLE
Support date/dateTime formatting

### PR DESCRIPTION
**Pull request by Janpot**
This PR proposes date formatting for the date and dateTime columns, in the flavour of the number formatter. Similarly it is built on top of `Intl.DateTimeFormat` and can be reused for other components just like we reuse the number format for the Metric component

Fixes https://github.com/mui/mui-toolpad/issues/2581

https://github.com/mui/mui-toolpad/assets/2109932/87a79166-5a93-4d77-8d9f-cca7586eb446

_edit_

Reversed the order since the video:

Screenshot 2023-08-31 at 15 15 50

### CODE REVIEW

File: packages/toolpad-core/src/dateFormat.tsx [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+);
+
+const TIME_FORMATS = new Map<DateStyle, Intl.DateTimeFormat>(
+  (['short', 'medium', 'long', 'full'] as const).map((timeStyle) => [
+    timeStyle,
+    new Intl.DateTimeFormat(undefined, { timeStyle }),
+  ]),
+);
+
+const DEMO_DATE = new Date();
```
Comment (apedroferreira) [inline packages/toolpad-core/src/dateFormat.tsx:42]: Could also be the current time? Anyway not important at all :D

**Comment by apedroferreira** [comment]
Just tried it, works great!
Not sure if it would be better for showing time (hours/minutes/seconds) to be an option all under the same column type, but it could depend on what obstacles that would cause or how much each of these types could be expanded.

File: packages/toolpad-core/src/dateFormat.tsx (hunk 1 above)
Comment (Janpot) [reply packages/toolpad-core/src/dateFormat.tsx:42]: I want a time that always demonstrates whether it's 12h or 24h clock

**Comment by Janpot** [comment]
> Not sure if it would be better for showing time (hours/minutes/seconds) to be an option all under the same column type, but it could depend on what obstacles that would cause or how much each of these types could be expanded.

You mean only having a `dateTime` type and not both `date` and `dateTime` types? I slightly agree with that, but the X grid has both so it could make sense to follow them for now. Switching to a single column type would be the same as deprecating the `date` type.

File: packages/toolpad-core/src/dateFormat.tsx (hunk 1 above)
Comment (Janpot) [reply packages/toolpad-core/src/dateFormat.tsx:43]: I've added a comment

**Comment by apedroferreira** [comment]
> > Not sure if it would be better for showing time (hours/minutes/seconds) to be an option all under the same column type, but it could depend on what obstacles that would cause or how much each of these types could be expanded.
> 
> You mean only having a `dateTime` type and not both `date` and `dateTime` types? I slightly agree with that, but the X grid has both so it could make sense to follow them for now. Switching to a single column type would be the same as deprecating the `date` type.

Ok, if the X grid separates them I guess we should do the same.

**Review by apedroferreira — APPROVED** [review]

**Pull request merged**